### PR TITLE
fix nested types generate

### PR DESCRIPTION
**Pull request by lifeengines**
fix for nested type generate

### CODE REVIEW

File: Il2CppInspector.Common/Reflection/TypeInfo.cs [hunk 1]
```diff
@@ -294,6 +294,8 @@ public class TypeArgumentsComparer : EqualityComparer<TypeInfo[]>
             var index = name.IndexOf("`", StringComparison.Ordinal);
             if (index != -1)
                 name = name.Remove(index);
+            //get rid of compiler generate characters that's c# not support
+            name = name.Replace("<", "_").Replace(">", "_");
```
Comment (djkaty) [inline Il2CppInspector.Common/Reflection/TypeInfo.cs:298]: This still changes `UnmangledBaseName` which is not a C#-specific name property. Since every use of `unmangleName` except for the definition of `UnmangledBaseName` is used for a C#-specific name, however, I think the best solution is to just change `UnmangledBaseName` so it doesn't use `unmangleName()` anymore.

File: Il2CppInspector.Common/Reflection/MethodBase.cs [hunk 2]
```diff
@@ -110,6 +110,19 @@ public abstract class MethodBase : MemberInfo
...
+            get
+            {
+                var index = CSharpName.LastIndexOf(".");
+                return (index >= 0)
+                        //implement for interface may contains generic parameters
+                        ? (CSharpName.Substring(0, index + 1) + CSharpName.Substring(index + 1).Replace("<", "_").Replace(">", "_"))
+                        : CSharpName.Replace("<", "_").Replace(">", "_");
+
+            }
+        }
```
Comment (djkaty) [inline Il2CppInspector.Common/Reflection/MethodBase.cs:125]: There is no need to create a new property here as we will end up with `CSharpName` and `CSharpSafeName`, we only need one property to correctly return the C# name of a method because there is only one possible correct name for a given method. You can add this code to the end of `MethodBase.CSharpName`.

But, will this not mess up generic method names, eg `void Foo<T>(T bla)`?

**Comment by djkaty** [comment]
The < and > need to be removed before any generic parameters are added to the type name. If it's easier for you I can try to figure out a way to do this later this afternoon.

**Comment by lifeengines** [comment]
I didn't find a proper way to remove <> before generic parameters.so I use a regular express to escape all the name 
for temporary use:
```
        private readonly static Regex rgxCGCharsReplace = new Regex(@"(^|\.)<(\w*?)>");
        public static string EscapeCSharpName(string name)
        {
            return (name.IndexOf("<") >= 0) ? rgxCGCharsReplace.Replace(name, "$1_$2_") : name;
        }
```
I assume all the compiler generator name are start with `<`,hope that's the real.
waiting for your update.thanks.

**Comment by djkaty** [comment]
Hey,

Sorry for not getting back to you sooner, this actually turned out to be a rather complex problem and it took me 2 straight days to mostly solve (getting it to work with one project is not too hard, the total test suite is currently 104 applications with many edge cases). I have made a number of commits, could you check out the latest build and let me know if it works correctly for you?

**Comment by lifeengines** [comment]
thanks very much.I will test immediately.

**Comment by lifeengines** [comment]
I do the test with my project,works very well.thanks again.this could be closed.

**Comment by djkaty** [comment]
Awesome, thanks!
